lmdeploy/pytorch/engine/logits_process.py

```diff
@@ -216,6 +222,24 @@ def to_device(self, device: str):
         return SamplingInputs(**out_dict)
 
 
+def _apply_custom_logits_processors(batched_logits_processors, input_ids,
+                                    seq_length, history_ids, logits):
+    """Apply custom logits processors."""
+    for seq_id, processors in enumerate(batched_logits_processors):
+        if processors is not None:
+            if history_ids.shape[-1] == 0:
+                seq_cumsum = [0] + seq_length.cumsum(0).tolist()
+                input_id = input_ids[0][seq_cumsum[seq_id]:seq_cumsum[seq_id +
+                                                                      1]]
+                for processor in processors:
+                    logits[seq_id] = processor(input_id, logits[seq_id])
+            else:
+                for processor in processors:
+                    logits[seq_id] = processor(history_ids[seq_id],
```

Comment:
what is the first input, input_ids or history_ids?

Reply:
It's just `function(previous_token_ids, logits)`. I wanted to align with vllm.
https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/logits_processor.py#L137

But I must pass more arguments in if I want to get `prompt_tokens_ids` and `past_tokens_ids`.

Reply:
Should we change history_ids to all_ids? That would also benefit replicate penalty.